### PR TITLE
Support for writing data from DynamoDB Streams to Cloud Spanner

### PR DESCRIPTION
**Pull request by nareshz**
Things added
- Converting data read from DynamoDB Streams to Cloud Spanner data based on source and spanner schemas.
- Creating mutations from converted data and processing them to Cloud Spanner.
- Collecting of bad and dropped records if errors are faced.
- Retry logic if insertion fails because of missing parent data.
- Updating customers about the progress of processing every minute. 
- Unit tests.

Things not handled
- Changes to report, badData files and Integration tests.
- Appropriate changes to README for using streaming migration.
- These will be added in subsequent PRs.

### CODE REVIEW

File: sources/dynamodb/streaming.go [hunk 1]
```diff
@@ -253,9 +315,133 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
...
+	srcSchema := conv.SrcSchema[srcTable]
+	spTable, err1 := internal.GetSpannerTable(conv, srcTable)
+	spCols, err2 := internal.GetSpannerCols(conv, srcTable, srcSchema.ColNames)
+	spSchema, ok := conv.SpSchema[spTable]
+	var err error
+	if err1 != nil || err2 != nil || !ok {
+		err = fmt.Errorf(fmt.Sprintf("err1=%s, err2=%s, ok=%t", err1, err2, ok))
+	}
+	return srcSchema, spTable, spCols, spSchema, err
+}
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:329]: This seems to be a generic method not specific to dynamodb so its better to place it in utility

File: sources/dynamodb/streaming.go [hunk 2]
```diff
@@ -253,9 +315,133 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
...
+	if len(badCols) == 0 {
+		writeRecord(streamInfo, srcTable, spTable, eventName, spCols, spVals, srcSchema)
+	} else {
+		streamInfo.StatsAddBadRecord(srcTable, eventName)
+		streamInfo.CollectBadRecord(eventName, srcTable, srcSchema.ColNames, srcStrVals)
+	}
 	streamInfo.StatsAddRecordProcessed()
 }
+
+// writeRecord checks if the writer to write data to Cloud Spanner is configured or not.
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:361]: Please rephrase this to make it clearer

File: sources/dynamodb/streaming.go [hunk 3]
```diff
@@ -253,9 +315,133 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
...
+		time.Sleep(4 * time.Second)
+		retryLimit--
+	}
+	return err
+}
+
+// setWriter initializes the write function used to write mutations to Cloud Spanner.
+func setWriter(streamInfo *StreamingInfo, client *sp.Client, conv *internal.Conv) {
+	streamInfo.write = func(m *sp.Mutation) error {
+		const migrationMetadataKey = "cloud-spanner-migration-metadata"
```
Comment (shreyakhajanchi) [inline sources/dynamodb/streaming.go:440]: Since this const is used at multiple places it would be better to put it in constants file and use it from there

**Review by shreyakhajanchi — COMMENTED** [review]
Added few comments

File: sources/dynamodb/streaming.go [hunk 4]
```diff
@@ -80,6 +90,58 @@ func catchCtrlC(wg *sync.WaitGroup, streamInfo *StreamingInfo) {
...
+
+		lastFiveMin += arr[counter]
+
+		if timer < 5 {
+			firstFiveMin += arr[counter]
+		}
+
+		lastMin := arr[counter]
+		optimumCondition := ((lastFiveMin*100 <= 5*firstFiveMin) || (lastMin == 0))
+		updateProgress(optimumCondition, false, tillLastMin)
```
[the remark this answers is not in the document]
Comment (nareshz) [reply sources/dynamodb/streaming.go:140]: streamInfo.recordsProcessed may increase during the time we are checking optimumCondition and calling updateProgress.
We are using tillLastMin to show that the last time we checked optimumCondition this was the number of records processed.

File: sources/dynamodb/streaming.go (hunk 1 above)
Comment (nareshz) [reply sources/dynamodb/streaming.go:329]: Okay! will do that.

File: sources/dynamodb/streaming.go [hunk 5]
```diff
@@ -253,9 +315,133 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
...
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0]})
+	} else {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0], reqSpVals[1]})
+	}
+	return m
+}
+
+// parentDataMissingError checks if the error is a parent data missing error.
+func parentDataMissingError(err error) bool {
+	return strings.Contains(err.Error(), "NotFound") && strings.Contains(err.Error(), "Parent row") && strings.Contains(err.Error(), "is missing")
```
[the remark this answers is not in the document]
Comment (nareshz) [reply sources/dynamodb/streaming.go:418]: I checked after our discussion and found that Contains is better in terms of performance.

File: sources/dynamodb/streaming.go [hunk 6]
```diff
@@ -407,9 +407,9 @@ func parentDataMissingError(err error) bool {
 }
 
 // writeMutation handles writing of a mutation to Cloud Spanner. If insertion fails
-// because of parent data missing error then it retries for a limit of 1000.
+// because of parent data missing error then it retries for a limit of 100.
 func writeMutation(m *sp.Mutation, streamInfo *StreamingInfo) error {
-	retryLimit := 1000
+	retryLimit := 100
```
Comment (manitgupta) [inline sources/dynamodb/streaming.go:412]: Don't we have a constants or config file in harbourbridge where we can declare these constants?
If not, lets declare `retryLimit` as a constant instead of a variable.

File: conversion/conversion.go [hunk 7]
```diff
@@ -329,7 +327,7 @@ func populateDataConv(conv *internal.Conv, config writer.BatchWriterConfig, clie
 		migrationData := metrics.GetMigrationData(conv, "", "", constants.DataConv)
 		serializedMigrationData, _ := proto.Marshal(migrationData)
 		migrationMetadataValue := base64.StdEncoding.EncodeToString(serializedMigrationData)
-		_, err := client.Apply(metadata.AppendToOutgoingContext(context.Background(), migrationMetadataKey, migrationMetadataValue), m)
+		_, err := client.Apply(metadata.AppendToOutgoingContext(context.Background(), constants.MigrationMetadataKey, migrationMetadataValue), m)
```
Comment (manitgupta) [inline conversion/conversion.go:330]: Is this change backward compatible? Can it cause a split with the existing migration usage data that we have/and or break the migration dashboard we have created?

File: sources/dynamodb/streaming.go (hunk 6 above)
Comment (nareshz) [reply sources/dynamodb/streaming.go:412]: we have a constant file but this retryLimit is very specific to DynamoDB. So it wouldn't make sense to put in constants file.

Okay! will change retryLimit to a constant.

File: sources/dynamodb/streaming.go [hunk 8]
```diff
@@ -253,9 +316,119 @@ func getRecords(streamClient dynamodbstreamsiface.DynamoDBStreamsAPI, shardItera
...
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0]})
+	} else {
+		m = sp.Delete(spTable, sp.Key{reqSpVals[0], reqSpVals[1]})
+	}
+	return m
+}
+
+// parentDataMissingError checks if the error is a parent data missing error.
+func parentDataMissingError(err error) bool {
+	return strings.Contains(err.Error(), "NotFound") && strings.Contains(err.Error(), "Parent row") && strings.Contains(err.Error(), "is missing")
```
Comment (manitgupta) [inline sources/dynamodb/streaming.go:406]: This is a pretty brittle piece of code. Depending on a the text of an error is risky. Unlike err types which are not usually change without notice to the client, the error text may change without notice.
If there is no way around this at this time, let's add a comment where on why we did this and create an issue in the repo to check we can come back and fix it later.

Comment (nareshz) [reply sources/dynamodb/streaming.go:406]: As this error is very specific in this solution, adding a comment on why we are doing this and a note that this may be subject to change.
